xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/DefaultMentionsDataConsumer.java

```diff
@@ -365,32 +418,66 @@ private void handleBaseObjectOnUpdate(List<BaseObject> oldEntry, BaseObject base
      * @param location the location of the property
      * @param documentReference the reference of the document holding the property
      * @param authorReference the reference of the author change
      * @param syntax the syntax of the document
+     * @param mentionNotificationParameters hold the list of identified mentions
      */
     private void handleProperty(Optional<BaseObject> oldBaseObject, LargeStringProperty largeStringProperty,
         MentionLocation location,
-        DocumentReference documentReference, DocumentReference authorReference, Syntax syntax)
+        DocumentReference documentReference, DocumentReference authorReference, Syntax syntax,
+        List<MentionNotificationParameters> mentionNotificationParameters)
     {
         Optional<XDOM> oldDom = oldBaseObject.flatMap(it -> ofNullable(it.getField(largeStringProperty.getName())))
-                                    .filter(it -> it instanceof LargeStringProperty)
-                                    .flatMap(
-                                        it -> this.xdomService.parse(((LargeStringProperty) it).getValue(), syntax));
+            .filter(it -> it instanceof LargeStringProperty)
+            .flatMap(
+                it -> this.xdomService.parse(((LargeStringProperty) it).getValue(), syntax));
         this.xdomService.parse(largeStringProperty.getValue(), syntax).ifPresent(xdom -> {
-            // can be replaced by ifPresentOrElse for in java 9+ 
-            oldDom.ifPresent(od -> handleUpdatedContent(od, xdom, documentReference, authorReference, location));
+            // can be replaced by ifPresentOrElse for in java 9+
+            oldDom.ifPresent(od -> handleUpdatedContent(od, xdom, documentReference, authorReference, location,
+                mentionNotificationParameters));
             if (!oldDom.isPresent()) {
-                handleCreatedContent(xdom, documentReference, authorReference, location);
+                handleCreatedContent(xdom, documentReference, authorReference, location, mentionNotificationParameters);
             }
         });
     }
 
-    private void sendNotification(MentionNotificationParameters notificationParameters)
+    private void sendNotification(List<MentionNotificationParameters> notificationParametersList)
+    {
+        for (MentionNotificationParameters mentionNotificationParameters : notificationParametersList) {
+
+            // If new types are are found, that means new mentions have been found in the entity.
+            // In this case, we retrieved the mention notification services for the type of users mentioned and call
+            // them with the list of mentions analyzed.
+            for (String type : mentionNotificationParameters.getNewTypes()) {
+                this.mentionNotificationServiceProvider.get(type)
+                    .ifPresent(it -> {
+                        try {
+                            it.sendNotifications(mentionNotificationParameters);
```

Comment:
So you actually send back all the mentions informations, for any services. Meaning if in the same comment I have one Mention to an XWiki user, and one Mention to an AP user, both services will be called with both mentions info? Doesn't sound really right.

Reply:
> For instance, the ActivityPub implementation will want to convert local actors to their activity pub identifier to add them to the mentioned actors of the entity, event if they have been mentioned as local actors.

The usecase is interesting, I'm not yet entirely sure to agree with it though. Sounds almost like an unwanted side-effect.

Reply:
The mentions of an entity are still sent together but we now rely on the existing event mechanism to notify listeners of the new mentions.